### PR TITLE
add fixed_toolbar_container_element setting which accepts an HTMLElement instead of a CSS3 selector string

### LINKED ISSUE #6559: add fixed_toolbar_container_element setting which accepts an HTMLElement instead of a CSS3 selector string
**Is your feature request related to a problem? Please describe.**
For an inline editor, you can currently only pass the `fixed_toolbar_container` CSS3 selector string to fix the toolbar to a certain div. However, when working with React Portals on different window (actually different body) instances, it might occur that the CSS3 selector fails to find the element on its current body.
Also, lately it's more common to pass HTMLElements instances around instead of selector strings.

**Describe the solution you'd like**
add the setting `fixed_toolbar_container_element` which accepts an HTMLElement instance in which the toolbar should be rendered. This should have a higher priority than `fixed_toolbar_container`.

**Describe alternatives you've considered**
/

### PR DESCRIPTION
**Pull request by pvrobays**
Related Ticket:
#6559 

Description of Changes:
* added setting `fixed_toolbar_container_element` which accepts an HTMLElement
* when `fixed_toolbar_container_element` is correctly passed, it will be used and `fixed_toolbar_container` will be ignored

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)
* [x] License headers added on new files (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):
Fixes #6559

### CODE REVIEW

**Comment by pvrobays** [comment]
closes #6559

**Comment by alecpl** [comment]
Does it have to be a separate option? Why not just support both cases via `fixed_toolbar_container`?

**Comment by pvrobays** [comment]
@alecpl I've put it in a separate option as I didn't want to break anything. `fixedContainerSelector` is currently hard coded to accept strings only.
But if you prefer to have it in the same option; that's possible of course. Just let me know and I'll adapt it.

**Comment by metricjs** [comment]
Hi @pvrobays,

First, thanks for your detailed explanation on the ticket. We do still have some concerns about the side-effects of using this in your environment, but in general we agree this is good functionality to add.

Adding this functionality as a separate setting is preferred, partly for the reasons you mentioned. The only change Tiny would like to request is renaming the setting to `fixed_toolbar_container_target`, to make it consistent with other similar settings. 

This PR will also need tests before we can merge it - are you happy to add those?

**Comment by alecpl** [comment]
It would be even more consistent with the rest if the same priority is used. I.e. `selector` option has precedence over `target`.

That being said. I still think it would be better to accept both element and selector in the same option, but it's not me to decide.

It would be definitely easier for the end-user to find this in documentation.

Let's take `selector` and `target` options as an example. `selector` is used in most examples in the documentation, so finding that there's also the `target` option is not easy, until you know the option name or read the whole documentation. So, the fact that a precedent exists should not be taken without a grain of salt.

**Comment by pvrobays** [comment]
Hi @metricjs 

I'll rename the setting to `fixed_toolbar_container_target`. If however you agree with @alecpl just let me know, I'm willing to adapt it. There's something to say for both solutions.

Concerning the tests, I'm happy to add some. I'm new to your testing environment though, but I'll give it a try :)

Could you also share some of your main concerns about the way we're currently using it in our environment? I'm curious; we might not have tested everything so far.

**Comment by pvrobays** [comment]
@alecpl Your comment about the priority of the settings is correct. I'll adapt it to have the selector setting priority over the target setting.

**Comment by pvrobays** [comment]
@metricjs the following changes done:

- rename the setting to `fixed_toolbar_container_target`
- change the priority of the settings to first try and use `fixed_toolbar_container` before `fixed_toolbar_container_target`
- added a few tests (they're copied and adapted, but I think they should be sufficient)

Please let me know if anything else should be done to get the PR merged.

**Comment by metricjs** [comment]
Thanks @pvrobays! I'll get the team to review this PR properly now. Hopefully that's everything.

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts
Comment (JamesToohey) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts:30]: This test can be a lot simpler because we only need to check that the toolbar exists, something like:

```
    const editor = hook.editor();
    await TinyUiActions.pWaitForUi(editor, '#toolbar .tox-editor-header');
    Assertions.assertPresence('Check that only one toolbar exists', { '.tox-editor-header': 1 }, SugarBody.body());
```

`TinyUiActions` is from `@ephox/mcagar`
`Assertions` is from `@ephox/agar`

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts
Comment (JamesToohey) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts:9]:
```suggestion
describe('browser.tinymce.themes.silver.editor.SilverFixedToolbarContainerPriorityTest', () => {
```

**Review by JamesToohey — COMMENTED** [review]
It's looking good! Thanks @pvrobays 

Well done jumping in and writing some tests, we know it is difficult.

I've added a few comments and suggestions for improvements. Please feel free to ask any questions about what I've said.

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts
Comment (pvrobays) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts:9]: Woops. Yeah I definitely didn't copy this from `SilverFixedToolbarContainerTest` 🙄

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts
Comment (pvrobays) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts:22]: added a test for this

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts
Comment (pvrobays) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts]: Awesome, thanks for the example.
The tests have been adapted.

**Review by pvrobays — COMMENTED** [review]
Thanks @JamesToohey for the quick response and the helpful input. 

All comments are addressed and should be fixed. Please let me know if there's anything else.

**Comment by pvrobays** [comment]
@metricjs Any update on when this might get pulled & released?

**Review by lnewson — APPROVED** [review]
Thanks for the addition @pvrobays! I've just left a couple comments for some small things that need fixing, but otherwise LGTM :+1:

File: modules/tinymce/src/themes/silver/main/ts/api/Settings.ts
Comment (pvrobays) [reply modules/tinymce/src/themes/silver/main/ts/api/Settings.ts:103]: Yeah, I like your specific typing. Too bad you can't pass some kind of generic type to `getParam<>()`

**Review by pvrobays — COMMENTED** [review]
@lnewson Thanks for the review. All suggestions were applied.

**Review by JamesToohey — APPROVED** [review]

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts
Comment (ltrouton) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts:18]: I think `toolbarWithId` needs to be removed here as well

**Review by ltrouton — APPROVED** [review]

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts
Comment (pvrobays) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/SilverFixedToolbarContainerPriorityTest.ts:18]: Yeah indeed. Nice find!

**Review by pvrobays — COMMENTED** [review]
Thanks for the review @ltrouton. 
The comment has been resolved.

**Comment by lnewson** [comment]
Hi @pvrobays, just to keep you in the loop, we're going to merge this into a local branch just so we can run the full range of our CI tests. Once that's successfully passed, we'll go ahead and merge it into the `develop` branch.

Thanks again for this contribution!

**Comment by lnewson** [comment]
Update: All tests passed, so it's now been merged to the `develop` branch :tada:

**Comment by pvrobays** [comment]
Great news @lnewson !
Any idea on when this might get released?

**Comment by metricjs** [comment]
@pvrobays This will be released with 5.8.0 - the Community release should be in early May, and the Enterprise release a couple weeks after that.

**Pull request merged**
